File: backend/api/agreementEndpoint.go
package api

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"superQiMiniAppBackend/alipay"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

// =========================================================================
// REQUEST/RESPONSE TYPES
// =========================================================================

type prepareContractRequest struct {
	ContractDescription string `json:"contractDescription" validate:"required"`
}

type applyAccessTokenRequest struct {
	AuthCode string `json:"authCode" validate:"required"`
}

type executeAgreementPaymentRequest struct {
	AccessToken      string `json:"accessToken" validate:"required"`
	Amount           int64  `json:"amount" validate:"required"`
	Currency         string `json:"currency"`
	OrderDescription string `json:"orderDescription"`
}

// =========================================================================
// ENDPOINT INITIALIZATION
// =========================================================================

func InitAgreementEndpoint(group fiber.Router) {
	log.Println("=================================================================")
	log.Println("[Backend] Initializing Agreement Payment Endpoints")
	log.Println("=================================================================")
	log.Println("[Backend] Registering route: POST /api/agreement/prepare")
	log.Println("[Backend] Registering route: POST /api/agreement/apply-token")
	log.Println("[Backend] Registering route: POST /api/agreement/pay")
	log.Println("=================================================================")

	agreementGroup := group.Group("/agreement")

	agreementGroup.Post("/prepare", handlePrepareContract)

	agreementGroup.Post("/apply-token", handleApplyAccessToken)

	agreementGroup.Post("/pay", handleExecuteAgreementPayment)
}

// =========================================================================
// STEP 1: PREPARE CONTRACT
// =========================================================================

func handlePrepareContract(ctx *fiber.Ctx) error {
	log.Println("\n=================================================================")
	log.Println("[Backend] INCOMING REQUEST: POST /api/agreement/prepare")
	log.Println("=================================================================")

	log.Println("[Backend] Request Headers:")
	ctx.Request().Header.VisitAll(func(key, value []byte) {
		log.Printf("[Backend]   %s: %s\n", string(key), string(value))
	})

	rawBody := string(ctx.Body())
	log.Printf("[Backend] Raw Request Body: %s\n", rawBody)

	var request prepareContractRequest
	if err := ctx.BodyParser(&request); err != nil {
		log.Printf("[Backend] ERROR: Failed to parse request body: %v\n", err)
		log.Println("=================================================================")
		return fiber.NewError(fiber.StatusBadRequest, "Invalid request body")
	}

	log.Printf("[Backend] SUCCESS: Request parsed successfully\n")
	log.Printf("[Backend] Contract description: %s\n", request.ContractDescription)
	log.Println("[Backend] -----------------------------------------------------------")
	log.Println("[Backend] Calling Alipay+ PrepareAuthorization API...")

	prepareResponse, err := alipay.Interface.PrepareAuthorization(request.ContractDescription)
	if err != nil {
		log.Printf("[Backend] ERROR: Alipay+ API call failed: %v\n", err)
		log.Println("[Backend] This could mean:")
		log.Println("[Backend]   1. Alipay+ gateway is unreachable")
		log.Println("[Backend]   2. Invalid credentials in .env file")
		log.Println("[Backend]   3. Network/firewall issue")
		log.Println("[Backend]   4. PrepareAuthorization function not implemented")
		log.Println("=================================================================")
		return fiber.NewError(fiber.StatusInternalServerError, "Failed to prepare contract: "+err.Error())
	}

	log.Println("[Backend] SUCCESS: Alipay+ API call successful")
	responseJSON, _ := json.MarshalIndent(prepareResponse, "", "  ")
	log.Printf("[Backend] Alipay+ Response:\n%s\n", string(responseJSON))
	log.Println("[Backend] -----------------------------------------------------------")

	log.Printf("[Backend] Checking result status: %s\n", prepareResponse.Result.ResultStatus)
	log.Printf("[Backend] Result code: %s\n", prepareResponse.Result.ResultCode)
	log.Printf("[Backend] Result message: %s\n", prepareResponse.Result.ResultMessage)

	if prepareResponse.Result.ResultStatus != "S" {
		log.Printf("[Backend] ERROR: Contract preparation failed\n")
		log.Printf("[Backend] Status: %s, Code: %s, Message: %s\n",
			prepareResponse.Result.ResultStatus,
			prepareResponse.Result.ResultCode,
			prepareResponse.Result.ResultMessage)
		log.Println("=================================================================")
		return ctx.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success":       false,
			"resultStatus":  prepareResponse.Result.ResultStatus,
			"resultCode":    prepareResponse.Result.ResultCode,
			"resultMessage": prepareResponse.Result.ResultMessage,
		})
	}

	if prepareResponse.AuthURL == "" {
		log.Printf("[Backend] WARNING: authUrl is EMPTY in response!\n")
		log.Printf("[Backend] This should not happen if resultStatus is 'S'\n")
		log.Printf("[Backend] Full response: %+v\n", prepareResponse)
		log.Println("=================================================================")
	} else {
		log.Printf("[Backend] SUCCESS: Authorization URL received: %s\n", prepareResponse.AuthURL)
	}

	response := fiber.Map{
		"success":       true,
		"authUrl":       prepareResponse.AuthURL,
		"resultStatus":  prepareResponse.Result.ResultStatus,
		"resultCode":    prepareResponse.Result.ResultCode,
		"resultMessage": prepareResponse.Result.ResultMessage,
	}

	log.Println("[Backend] -----------------------------------------------------------")
	log.Println("[Backend] Building response to frontend:")
	responseToFrontendJSON, _ := json.MarshalIndent(response, "", "  ")
	log.Printf("%s\n", string(responseToFrontendJSON))
	log.Println("[Backend] SUCCESS: Sending response to frontend")
	log.Println("=================================================================")

	return ctx.JSON(response)
}

// =========================================================================
// STEP 2: APPLY ACCESS TOKEN
// =========================================================================

func handleApplyAccessToken(ctx *fiber.Ctx) error {
	log.Println("\n=================================================================")
	log.Println("[Backend] INCOMING REQUEST: POST /api/agreement/apply-token")
	log.Println("=================================================================")

	rawBody := string(ctx.Body())
	log.Printf("[Backend] Raw Request Body: %s\n", rawBody)

	var request applyAccessTokenRequest
	if err := ctx.BodyParser(&request); err != nil {
		log.Printf("[Backend] ERROR: Failed to parse request body: %v\n", err)
		log.Println("=================================================================")
		return fiber.NewError(fiber.StatusBadRequest, "Invalid request body")
	}

	log.Printf("[Backend] SUCCESS: Request parsed successfully\n")
	log.Printf("[Backend] Auth code received (first 20 chars): %s...\n", truncateString(request.AuthCode, 20))
	log.Printf("[Backend] Auth code length: %d\n", len(request.AuthCode))
	log.Println("[Backend] -----------------------------------------------------------")
	log.Println("[Backend] Calling Alipay+ ApplyToken API...")

	tokenResponse, err := alipay.Interface.ApplyToken(request.AuthCode)
	if err != nil {
		log.Printf("[Backend] ERROR: Token exchange failed: %v\n", err)
		log.Println("=================================================================")
		return fiber.NewError(fiber.StatusBadRequest, "Token exchange failed: "+err.Error())
	}

	log.Println("[Backend] SUCCESS: Alipay+ API call successful")
	responseJSON, _ := json.MarshalIndent(tokenResponse, "", "  ")
	log.Printf("[Backend] Alipay+ Response:\n%s\n", string(responseJSON))
	log.Println("[Backend] -----------------------------------------------------------")

	log.Printf("[Backend] Checking result status: %s\n", tokenResponse.Result.ResultStatus)
	log.Printf("[Backend] Result code: %s\n", tokenResponse.Result.ResultCode)

	if tokenResponse.Result.ResultStatus != "S" || tokenResponse.Result.ResultCode != "SUCCESS" {
		log.Printf("[Backend] ERROR: Token application failed: %s\n", tokenResponse.Result.ResultMessage)
		log.Println("=================================================================")
		return ctx.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success":       false,
			"resultStatus":  tokenResponse.Result.ResultStatus,
			"resultCode":    tokenResponse.Result.ResultCode,
			"resultMessage": tokenResponse.Result.ResultMessage,
		})
	}

	log.Printf("[Backend] SUCCESS: Access token obtained\n")
	log.Printf("[Backend] Access Token (first 20 chars): %s...\n", truncateString(tokenResponse.AccessToken, 20))
	log.Printf("[Backend] Refresh Token (first 20 chars): %s...\n", truncateString(tokenResponse.RefreshToken, 20))
	log.Printf("[Backend] Token Expiry: %s\n", tokenResponse.AccessTokenExpiryTime)
	log.Printf("[Backend] Customer ID: %s\n", tokenResponse.CustomerID)
	log.Printf("[Backend] IMPORTANT: Store this accessToken in database for future payments!\n")
	log.Println("[Backend] SUCCESS: Sending response to frontend")
	log.Println("=================================================================")

	return ctx.JSON(fiber.Map{
		"success":                true,
		"accessToken":            tokenResponse.AccessToken,
		"refreshToken":           tokenResponse.RefreshToken,
		"accessTokenExpiryTime":  tokenResponse.AccessTokenExpiryTime,
		"refreshTokenExpiryTime": tokenResponse.RefreshTokenExpiryTime,
		"customerId":             tokenResponse.CustomerID,
		"resultStatus":           tokenResponse.Result.ResultStatus,
		"resultCode":             tokenResponse.Result.ResultCode,
		"resultMessage":          tokenResponse.Result.ResultMessage,
	})
}

// =========================================================================
// STEP 3: EXECUTE AGREEMENT PAYMENT
// =========================================================================

func handleExecuteAgreementPayment(ctx *fiber.Ctx) error {
	log.Println("\n=================================================================")
	log.Println("[Backend] INCOMING REQUEST: POST /api/agreement/pay")
	log.Println("=================================================================")

	rawBody := string(ctx.Body())
	log.Printf("[Backend] Raw Request Body: %s\n", rawBody)

	var request executeAgreementPaymentRequest
	if err := ctx.BodyParser(&request); err != nil {
		log.Printf("[Backend] ERROR: Failed to parse request body: %v\n", err)
		log.Println("=================================================================")
		return fiber.NewError(fiber.StatusBadRequest, "Invalid request body")
	}

	if request.Currency == "" {
		request.Currency = "IQD"
	}
	if request.OrderDescription == "" {
		request.OrderDescription = "Agreement payment - Monthly subscription"
	}

	log.Printf("[Backend] SUCCESS: Request parsed successfully\n")
	log.Printf("[Backend] Access Token (first 20 chars): %s...\n", truncateString(request.AccessToken, 20))
	log.Printf("[Backend] Amount: %d %s\n", request.Amount, request.Currency)
	log.Printf("[Backend] Order Description: %s\n", request.OrderDescription)
	log.Println("[Backend] -----------------------------------------------------------")

	if request.Amount <= 0 {
		log.Println("[Backend] ERROR: Invalid payment amount")
		log.Println("=================================================================")
		return ctx.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success":       false,
			"resultStatus":  "F",
			"resultMessage": "Payment amount must be greater than 0",
		})
	}

	if request.AccessToken == "" {
		log.Println("[Backend] ERROR: Access token is required")
		log.Println("=================================================================")
		return ctx.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success":       false,
			"resultStatus":  "F",
			"resultMessage": "Access token is required",
		})
	}

	log.Println("[Backend] Executing agreement payment...")

	paymentResponse, err := executeAgreementPaymentInternal(request.AccessToken, request.Amount, request.Currency, request.OrderDescription)
	if err != nil {
		log.Printf("[Backend] ERROR: Failed to execute payment: %v\n", err)
		log.Println("=================================================================")
		return ctx.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success":       false,
			"resultStatus":  "F",
			"resultMessage": err.Error(),
		})
	}

	response := buildAgreementPaymentResponse(paymentResponse)

	log.Println("[Backend] SUCCESS: Sending payment response to frontend")
	log.Println("=================================================================")
	return ctx.JSON(response)
}

// =========================================================================
// INTERNAL HELPER FUNCTIONS
// =========================================================================

func executeAgreementPaymentInternal(accessToken string, amount int64, currency string, orderDescription string) (alipay.PaymentResponse, error) {
	log.Println("[Backend] Preparing agreement payment request...")

	paymentRequestID := fmt.Sprintf("AGREEMENT-PAY-%s-%d", uuid.New().String(), time.Now().Unix())
	log.Printf("[Backend] Generated Payment Request ID: %s\n", paymentRequestID)

	expiryTime := time.Now().Add(30 * time.Minute).Format("2006-01-02T15:04:05-07:00")

	baseURL := os.Getenv("BASE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:1999"
	}

	paymentRequest := alipay.PaymentRequest{
		ProductCode:      alipay.AGREEMENT_PAYMENT, // Important: Use agreement payment product code
		PaymentRequestID: paymentRequestID,
		PaymentAuthCode:  accessToken, // This is the key difference - we use accessToken here
		PaymentAmount: alipay.PaymentAmount{
			Currency: currency,
			Value:    fmt.Sprintf("%d", amount),
		},
		Order: alipay.Order{
			OrderDescription: orderDescription,
		},
		PaymentExpiryTime: expiryTime,
		PaymentNotifyURL:  baseURL + "/api/webhook/payment-notify",
	}

	requestJSON, _ := json.MarshalIndent(paymentRequest, "", "  ")
	log.Printf("[Backend] Agreement payment request:\n%s\n", string(requestJSON))

	log.Println("[Backend] Calling /v1/payments/pay API...")
	paymentResponse, err := alipay.Interface.Pay(paymentRequest)
	if err != nil {
		log.Printf("[Backend] ERROR: Payment API call failed: %v\n", err)
		return alipay.PaymentResponse{}, fmt.Errorf("payment API call failed: %v", err)
	}

	responseJSON, _ := json.MarshalIndent(paymentResponse, "", "  ")
	log.Printf("[Backend] Payment API response:\n%s\n", string(responseJSON))

	switch paymentResponse.Result.ResultStatus {
	case "S":
		log.Println("[Backend] SUCCESS: Payment completed immediately")
		log.Printf("[Backend] Payment ID: %s\n", paymentResponse.PaymentID)
		log.Printf("[Backend] Payment Time: %s\n", paymentResponse.PaymentTime)
		log.Println("[Backend] WARNING: Money deducted from user's wallet automatically!")

	case "U":
		log.Println("[Backend] WARNING: Payment status unknown")
		log.Println("[Backend] Backend should poll /v1/payments/inquiryPayment for status")

	case "F":
		log.Printf("[Backend] ERROR: Payment failed - %s\n", paymentResponse.Result.ResultMessage)
		log.Printf("[Backend] Error Code: %s\n", paymentResponse.Result.ResultCode)

	default:
		log.Printf("[Backend] WARNING: Unexpected status: %s\n", paymentResponse.Result.ResultStatus)
	}

	return paymentResponse, nil
}

func buildAgreementPaymentResponse(paymentResponse alipay.PaymentResponse) fiber.Map {
	response := fiber.Map{
		"resultStatus":  paymentResponse.Result.ResultStatus,
		"resultCode":    paymentResponse.Result.ResultCode,
		"resultMessage": paymentResponse.Result.ResultMessage,
	}

	switch paymentResponse.Result.ResultStatus {
	case "S":
		response["status"] = "SUCCESS"
		response["success"] = true
		if paymentResponse.PaymentID != "" {
			response["paymentId"] = paymentResponse.PaymentID
		}
		if paymentResponse.PaymentRequestID != "" {
			response["paymentRequestId"] = paymentResponse.PaymentRequestID
		}
		if paymentResponse.PaymentTime != "" {
			response["paymentTime"] = paymentResponse.PaymentTime
		}

	case "U":
		response["status"] = "PENDING"
		response["success"] = false
		response["message"] = "Payment status is unknown. Backend should poll for status."
		if paymentResponse.PaymentID != "" {
			response["paymentId"] = paymentResponse.PaymentID
		}

	case "F":
		response["status"] = "FAILED"
		response["success"] = false
		response["message"] = paymentResponse.Result.ResultMessage

	default:
		response["status"] = "UNKNOWN"
		response["success"] = false
		response["message"] = "Unexpected payment status: " + paymentResponse.Result.ResultStatus
	}

	return response
}

func truncateString(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen]
}

File: backend-node/src/api/agreementEndpoint.js
const { v4: uuidv4 } = require('uuid');
const { prepareAuthorization, applyToken, pay, AGREEMENT_PAYMENT } = require('../alipay');

function initAgreementEndpoint(router) {
    console.log('=================================================================');
    console.log('[Backend] Initializing Agreement Payment Endpoints');
    console.log('=================================================================');
    console.log('[Backend] Registering route: POST /api/agreement/prepare');
    console.log('[Backend] Registering route: POST /api/agreement/apply-token');
    console.log('[Backend] Registering route: POST /api/agreement/pay');
    console.log('=================================================================');

    router.post('/agreement/prepare', handlePrepareContract);

    router.post('/agreement/apply-token', handleApplyAccessToken);

    router.post('/agreement/pay', handleExecuteAgreementPayment);
}

async function handlePrepareContract(req, res) {
    console.log('\n=================================================================');
    console.log('[Backend] INCOMING REQUEST: POST /api/agreement/prepare');
    console.log('=================================================================');

    console.log('[Backend] Request Headers:');
    Object.keys(req.headers).forEach(key => {
        console.log(`[Backend]   ${key}: ${req.headers[key]}`);
    });

    console.log(`[Backend] Raw Request Body: ${JSON.stringify(req.body)}`);

    const { contractDescription } = req.body;

    if (!contractDescription) {
        console.log('[Backend] ERROR: contractDescription is required');
        console.log('=================================================================');
        return res.status(400).json({ error: 'contractDescription is required' });
    }

    console.log('[Backend] SUCCESS: Request parsed successfully');
    console.log(`[Backend] Contract description: ${contractDescription}`);
    console.log('[Backend] -----------------------------------------------------------');
    console.log('[Backend] Calling Alipay+ PrepareAuthorization API...');

    try {
        const prepareResponse = await prepareAuthorization(contractDescription);

        console.log('[Backend] SUCCESS: Alipay+ API call successful');
        console.log('[Backend] Alipay+ Response:');
        console.log(JSON.stringify(prepareResponse, null, 2));
        console.log('[Backend] -----------------------------------------------------------');

        console.log(`[Backend] Checking result status: ${prepareResponse.result.resultStatus}`);
        console.log(`[Backend] Result code: ${prepareResponse.result.resultCode}`);
        console.log(`[Backend] Result message: ${prepareResponse.result.resultMessage}`);

        if (prepareResponse.result.resultStatus !== 'S') {
            console.log('[Backend] ERROR: Contract preparation failed');
            console.log(`[Backend] Status: ${prepareResponse.result.resultStatus}, Code: ${prepareResponse.result.resultCode}, Message: ${prepareResponse.result.resultMessage}`);
            console.log('=================================================================');
            return res.status(400).json({
                success: false,
                resultStatus: prepareResponse.result.resultStatus,
                resultCode: prepareResponse.result.resultCode,
                resultMessage: prepareResponse.result.resultMessage
            });
        }

        if (!prepareResponse.authUrl) {
            console.log('[Backend] WARNING: authUrl is EMPTY in response!');
            console.log('[Backend] This should not happen if resultStatus is \'S\'');
            console.log(`[Backend] Full response: ${JSON.stringify(prepareResponse)}`);
            console.log('=================================================================');
        } else {
            console.log(`[Backend] SUCCESS: Authorization URL received: ${prepareResponse.authUrl}`);
        }

        const response = {
            success: true,
            authUrl: prepareResponse.authUrl,
            resultStatus: prepareResponse.result.resultStatus,
            resultCode: prepareResponse.result.resultCode,
            resultMessage: prepareResponse.result.resultMessage
        };

        console.log('[Backend] -----------------------------------------------------------');
        console.log('[Backend] Building response to frontend:');
        console.log(JSON.stringify(response, null, 2));
        console.log('[Backend] SUCCESS: Sending response to frontend');
        console.log('=================================================================');

        return res.json(response);

    } catch (err) {
        console.log(`[Backend] ERROR: Alipay+ API call failed: ${err.message}`);
        console.log('[Backend] This could mean:');
        console.log('[Backend]   1. Alipay+ gateway is unreachable');
        console.log('[Backend]   2. Invalid credentials in .env file');
        console.log('[Backend]   3. Network/firewall issue');
        console.log('[Backend]   4. PrepareAuthorization function not implemented');
        console.log('=================================================================');
        return res.status(500).json({ error: 'Failed to prepare contract: ' + err.message });
    }
}

async function handleApplyAccessToken(req, res) {
    console.log('\n=================================================================');
    console.log('[Backend] INCOMING REQUEST: POST /api/agreement/apply-token');
    console.log('=================================================================');

    console.log(`[Backend] Raw Request Body: ${JSON.stringify(req.body)}`);

    const { authCode } = req.body;

    if (!authCode) {
        console.log('[Backend] ERROR: authCode is required');
        console.log('=================================================================');
        return res.status(400).json({ error: 'authCode is required' });
    }

    console.log('[Backend] SUCCESS: Request parsed successfully');
    console.log(`[Backend] Auth code received (first 20 chars): ${truncateString(authCode, 20)}...`);
    console.log(`[Backend] Auth code length: ${authCode.length}`);
    console.log('[Backend] -----------------------------------------------------------');
    console.log('[Backend] Calling Alipay+ ApplyToken API...');

    try {
        const tokenResponse = await applyToken(authCode);

        console.log('[Backend] SUCCESS: Alipay+ API call successful');
        console.log('[Backend] Alipay+ Response:');
        console.log(JSON.stringify(tokenResponse, null, 2));
        console.log('[Backend] -----------------------------------------------------------');

        console.log(`[Backend] Checking result status: ${tokenResponse.result.resultStatus}`);
        console.log(`[Backend] Result code: ${tokenResponse.result.resultCode}`);

        if (tokenResponse.result.resultStatus !== 'S' || tokenResponse.result.resultCode !== 'SUCCESS') {
            console.log(`[Backend] ERROR: Token application failed: ${tokenResponse.result.resultMessage}`);
            console.log('=================================================================');
            return res.status(400).json({
                success: false,
                resultStatus: tokenResponse.result.resultStatus,
                resultCode: tokenResponse.result.resultCode,
                resultMessage: tokenResponse.result.resultMessage
            });
        }

        console.log(`[Backend] Access Token (first 30 chars): ${truncateString(tokenResponse.accessToken, 30)}...`);
        console.log(`[Backend] Customer ID: ${tokenResponse.customerId}`);
        console.log(`[Backend] Access Token expiry: ${tokenResponse.accessTokenExpiryTime}`);

        const response = {
            success: true,
            accessToken: tokenResponse.accessToken,
            customerId: tokenResponse.customerId,
            accessTokenExpiryTime: tokenResponse.accessTokenExpiryTime,
            resultStatus: tokenResponse.result.resultStatus,
            resultCode: tokenResponse.result.resultCode,
            resultMessage: tokenResponse.result.resultMessage
        };

        console.log('[Backend] SUCCESS: Sending token response to frontend');
        console.log('=================================================================');

        return res.json(response);

    } catch (err) {
        console.log(`[Backend] ERROR: Token exchange failed: ${err.message}`);
        console.log('=================================================================');
        return res.status(400).json({ error: 'Token exchange failed: ' + err.message });
    }
}

async function handleExecuteAgreementPayment(req, res) {
    console.log('\n=================================================================');
    console.log('[Backend] INCOMING REQUEST: POST /api/agreement/pay');
    console.log('=================================================================');

    console.log(`[Backend] Raw Request Body: ${JSON.stringify(req.body)}`);

    let { accessToken, customerId, amount, currency, orderDescription } = req.body;

    if (!currency) {
        currency = 'IQD';
    }
    if (!orderDescription) {
        orderDescription = 'Agreement payment - Monthly subscription';
    }

    console.log('[Backend] SUCCESS: Request parsed successfully');
    console.log(`[Backend] Access Token (first 20 chars): ${truncateString(accessToken, 20)}...`);
    console.log(`[Backend] Customer ID: ${customerId}`);
    console.log(`[Backend] Amount: ${amount} ${currency}`);
    console.log(`[Backend] Order Description: ${orderDescription}`);
    console.log('[Backend] -----------------------------------------------------------');

    if (!amount || amount <= 0) {
        console.log('[Backend] ERROR: Invalid payment amount');
        console.log('=================================================================');
        return res.status(400).json({
            success: false,
            resultStatus: 'F',
            resultMessage: 'Payment amount must be greater than 0'
        });
    }

    if (!accessToken) {
        console.log('[Backend] ERROR: Access token is required');
        console.log('=================================================================');
        return res.status(400).json({
            success: false,
            resultStatus: 'F',
            resultMessage: 'Access token is required'
        });
    }

    if (!customerId) {
        console.log('[Backend] ERROR: Customer ID is required');
        console.log('=================================================================');
        return res.status(400).json({
            success: false,
            resultStatus: 'F',
            resultMessage: 'Customer ID is required'
        });
    }

    console.log('[Backend] Executing agreement payment...');

    try {
        const paymentResponse = await executeAgreementPaymentInternal(
            accessToken,
            customerId,
            amount,
            currency,
            orderDescription
        );

        const response = buildAgreementPaymentResponse(paymentResponse);

        console.log('[Backend] SUCCESS: Sending payment response to frontend');
        console.log('=================================================================');
        return res.json(response);

    } catch (err) {
        console.log(`[Backend] ERROR: Failed to execute payment: ${err.message}`);
        console.log('=================================================================');
        return res.status(500).json({
            success: false,
            resultStatus: 'F',
            resultMessage: err.message
        });
    }
}

async function executeAgreementPaymentInternal(accessToken, customerId, amount, currency, orderDescription) {
    console.log('[Backend] Preparing agreement payment request...');
    console.log(`[Backend] Using Customer ID: ${customerId}`);

    const paymentRequestId = `AGREEMENT-PAY-${uuidv4()}-${Date.now()}`;
    console.log(`[Backend] Generated Payment Request ID: ${paymentRequestId}`);

    const expiryTime = new Date(Date.now() + 30 * 60 * 1000)
        .toISOString()
        .replace('Z', '+00:00');

    const baseUrl = process.env.BASE_URL || 'http://localhost:1999';

    const paymentRequest = {
        productCode: AGREEMENT_PAYMENT,
        paymentRequestId: paymentRequestId,
        paymentAuthCode: accessToken,
        paymentAmount: {
            currency: currency,
            value: amount.toString()
        },
        order: {
            orderDescription: orderDescription,
            buyer: {
                referenceBuyerId: customerId
            }
        },
        paymentExpiryTime: expiryTime,
        paymentNotifyUrl: baseUrl + '/api/webhook/payment-notify'
    };

    console.log('[Backend] Agreement payment request:');
    console.log(JSON.stringify(paymentRequest, null, 2));

    console.log('[Backend] Calling /v1/payments/pay API...');
    const paymentResponse = await pay(paymentRequest);

    console.log('[Backend] Payment API response:');
    console.log(JSON.stringify(paymentResponse, null, 2));

    switch (paymentResponse.result.resultStatus) {
        case 'S':
            console.log('[Backend] SUCCESS: Payment completed immediately');
            console.log(`[Backend] Payment ID: ${paymentResponse.paymentId}`);
            console.log(`[Backend] Payment Time: ${paymentResponse.paymentTime}`);
            console.log('[Backend] Money deducted from user\'s wallet automatically!');
            break;

        case 'U':
            console.log('[Backend] WARNING: Payment status unknown');
            console.log('[Backend] Backend should poll /v1/payments/inquiryPayment for status');
            break;

        case 'F':
            console.log(`[Backend] ERROR: Payment failed - ${paymentResponse.result.resultMessage}`);
            console.log(`[Backend] Error Code: ${paymentResponse.result.resultCode}`);
            break;

        default:
            console.log(`[Backend] WARNING: Unexpected status: ${paymentResponse.result.resultStatus}`);
    }

    return paymentResponse;
}

function buildAgreementPaymentResponse(paymentResponse) {
    const response = {
        resultStatus: paymentResponse.result.resultStatus,
        resultCode: paymentResponse.result.resultCode,
        resultMessage: paymentResponse.result.resultMessage
    };

    switch (paymentResponse.result.resultStatus) {
        case 'S':
            response.status = 'SUCCESS';
            response.success = true;
            if (paymentResponse.paymentId) {
                response.paymentId = paymentResponse.paymentId;
            }
            if (paymentResponse.paymentRequestId) {
                response.paymentRequestId = paymentResponse.paymentRequestId;
            }
            if (paymentResponse.paymentTime) {
                response.paymentTime = paymentResponse.paymentTime;
            }
            break;

        case 'U':
            response.status = 'PENDING';
            response.success = false;
            response.message = 'Payment status is unknown. Backend should poll for status.';
            if (paymentResponse.paymentId) {
                response.paymentId = paymentResponse.paymentId;
            }
            break;

        case 'F':
            response.status = 'FAILED';
            response.success = false;
            response.message = paymentResponse.result.resultMessage;
            break;

        default:
            response.status = 'UNKNOWN';
            response.success = false;
            response.message = 'Unexpected payment status: ' + paymentResponse.result.resultStatus;
    }

    return response;
}

function truncateString(s, maxLen) {
    if (!s) return '';
    if (s.length <= maxLen) {
        return s;
    }
    return s.substring(0, maxLen);
}

module.exports = { initAgreementEndpoint };